File: Darts/src/burlton/dartzee/code/screen/PlayerSelectDialog.java
package burlton.dartzee.code.screen;

import burlton.core.code.obj.HandyArrayList;
import burlton.dartzee.code.bean.PlayerTypeFilterPanel;
import burlton.dartzee.code.bean.ScrollTablePlayers;
import burlton.dartzee.code.db.PlayerEntity;
import burlton.desktopcore.code.screen.SimpleDialog;
import burlton.desktopcore.code.util.DialogUtil;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.util.ArrayList;
import java.util.List;

public class PlayerSelectDialog extends SimpleDialog
{
	private List<PlayerEntity> players = new ArrayList<>();
	private HandyArrayList<PlayerEntity> selectedPlayers = new HandyArrayList<>();
	private ArrayList<PlayerEntity> playersToExclude = new ArrayList<>();
	
	public PlayerSelectDialog(int selectionMode) 
	{
		super();
		setTitle("Select Player(s)");
		setSize(300, 300);
		setModal(true);
		
		getContentPane().add(panelNorth, BorderLayout.NORTH);
		getContentPane().add(tablePlayers, BorderLayout.CENTER);
		tablePlayers.setSelectionMode(selectionMode);
		
		panelNorth.addActionListener(this);
	}
	
	private final PlayerTypeFilterPanel panelNorth = new PlayerTypeFilterPanel();
	private final ScrollTablePlayers tablePlayers = new ScrollTablePlayers();
	
	@Override
	public void actionPerformed(ActionEvent arg0) 
	{
		if (panelNorth.isEventSource(arg0))
		{
			init();
		}
		else
		{
			super.actionPerformed(arg0);
		}
	}
	
	private void init()
	{
		retrievePlayers();
		populateTable();
	}
	
	private void retrievePlayers()
	{
		String whereSql = panelNorth.getWhereSql();
		players = PlayerEntity.retrievePlayers(whereSql, false);
		
		for (int i=players.size()-1; i >= 0; i--)
		{
			PlayerEntity player = players.get(i);
			if (playersToExclude.contains(player))
			{
				players.remove(i);
			}
		}
	}
	
	private void populateTable()
	{
		tablePlayers.initTableModel(players);
	}
	
	/**
	 * Gets / sets
	 */
	public void setPlayersToExclude(ArrayList<PlayerEntity> playersToExclude)
	{
		this.playersToExclude = playersToExclude;
	}
	public HandyArrayList<PlayerEntity> getSelectedPlayers()
	{
		return selectedPlayers;
	}
	
	/**
	 * Statics
	 */
	public static PlayerEntity selectPlayer()
	{
		HandyArrayList<PlayerEntity> players = selectPlayers(new ArrayList<>(), ListSelectionModel.SINGLE_SELECTION);
		if (players.isEmpty())
		{
			return null;
		}
		
		return players.firstElement();
	}
	public static HandyArrayList<PlayerEntity> selectPlayers(ArrayList<PlayerEntity> playersToExclude)
	{
		return selectPlayers(playersToExclude, ListSelectionModel.MULTIPLE_INTERVAL_SELECTION);
	}
	private static HandyArrayList<PlayerEntity> selectPlayers(ArrayList<PlayerEntity> playersToExclude, int selectionMode)
	{
		PlayerSelectDialog dialog = new PlayerSelectDialog(selectionMode);
		dialog.setPlayersToExclude(playersToExclude);
		dialog.init();
		dialog.setLocationRelativeTo(null);
		dialog.setVisible(true);
		
		return dialog.getSelectedPlayers();
	}

	@Override
	public void okPressed()
	{
		selectedPlayers = tablePlayers.getSelectedPlayers();
		if (selectedPlayers.isEmpty())
		{
			DialogUtil.showError("You must select at least one player.");
			return;
		}
		
		dispose();
	}
}

File: Darts/src/burlton/dartzee/code/screen/PlayerSelectDialog.kt
package burlton.dartzee.code.screen

import burlton.dartzee.code.bean.PlayerTypeFilterPanel
import burlton.dartzee.code.bean.getSelectedPlayers
import burlton.dartzee.code.bean.initPlayerTableModel
import burlton.dartzee.code.db.PlayerEntity
import burlton.dartzee.code.core.bean.ScrollTable
import burlton.desktopcore.code.screen.SimpleDialog
import burlton.desktopcore.code.util.DialogUtil
import java.awt.BorderLayout
import java.awt.event.ActionEvent
import javax.swing.ListSelectionModel

class PlayerSelectDialog(selectionMode: Int) : SimpleDialog()
{
    var selectedPlayers = listOf<PlayerEntity>()
    var playersToExclude = listOf<PlayerEntity>()

    val panelNorth = PlayerTypeFilterPanel()
    val tablePlayers = ScrollTable()

    init
    {
        title = "Select Player(s)"
        setSize(300, 300)
        isModal = true

        contentPane.add(panelNorth, BorderLayout.NORTH)
        contentPane.add(tablePlayers, BorderLayout.CENTER)
        tablePlayers.setSelectionMode(selectionMode)

        panelNorth.addActionListener(this)
    }

    override fun actionPerformed(arg0: ActionEvent)
    {
        if (panelNorth.isEventSource(arg0))
        {
            buildTable()
        }
        else
        {
            super.actionPerformed(arg0)
        }
    }

    fun buildTable()
    {
        val whereSql = panelNorth.getWhereSql()
        val allPlayers = PlayerEntity.retrievePlayers(whereSql, false)

        val players = allPlayers.filter{ p -> playersToExclude.none{ it.rowId == p.rowId} }
        tablePlayers.initPlayerTableModel(players)
    }

    override fun okPressed()
    {
        selectedPlayers = tablePlayers.getSelectedPlayers()
        if (selectedPlayers.isEmpty())
        {
            DialogUtil.showError("You must select at least one player.")
            return
        }

        dispose()
    }

    companion object
    {
        fun selectPlayer(): PlayerEntity?
        {
            val players = selectPlayers(listOf(), ListSelectionModel.SINGLE_SELECTION)
            return if (players.isEmpty()) null else players.first()
        }

        fun selectPlayers(playersToExclude: List<PlayerEntity>): List<PlayerEntity>
        {
            return selectPlayers(playersToExclude, ListSelectionModel.MULTIPLE_INTERVAL_SELECTION)
        }

        private fun selectPlayers(playersToExclude: List<PlayerEntity>, selectionMode: Int): List<PlayerEntity>
        {
            val dialog = PlayerSelectDialog(selectionMode)
            dialog.playersToExclude = playersToExclude
            dialog.buildTable()
            dialog.setLocationRelativeTo(null)
            dialog.isVisible = true

            return dialog.selectedPlayers
        }
    }
}
